lib/src/result.dart

```diff
@@ -1,27 +1,44 @@
 class Result<T> {
-  final List<dynamic> results;
+  final List<T> results;
   final int total;
   final int limit;
   final int offset;
 
-  Result(
-      {this.results = const [],
-      this.limit = 0,
-      this.offset = 0,
-      this.total = 0});
+  const Result({
+    this.results = const [],
+    this.limit = 0,
+    this.offset = 0,
+    this.total = 0,
+  });
 
-  factory Result.fromMapWithType(Map<String, dynamic> map, fromMap) =>
+  factory Result.fromMapWithType(
+    Map<String, Object?> map,
+    T Function(Map<String, Object?> item) fromMap,
+  ) =>
       Result<T>(
-        results: map['results'].map((e) => fromMap(e)).toList(),
+        results: (map['results'] as Iterable?)
+                ?.cast<Map<String, Object?>>()
+                .map((e) => fromMap(e))
+                .toList() ??
+            [],
         total: map['total'] as int,
         offset: map['offset'] as int,
         limit: map['limit'] as int,
       );
 
-  factory Result.fromMap(Map<String, dynamic> map) => Result(
-        results: map['results'] as List<T>,
+  factory Result.fromMap(Map<String, Object?> map) => Result(
+        results: (map['results'] as Iterable?)?.cast<T>().toList() ?? [],
         total: map['total'] as int,
         offset: map['offset'] as int,
         limit: map['limit'] as int,
       );
+
+  Result<TTarget> map<TTarget>(TTarget Function(T item) converter) {
```

Comment:
well, we can follow `firestore` conventions by renaming it to `withConverter` instead of just `map`, but I made this public intentionally because this is a very common class used by every application